src/frontend/src/binder/expr/function.rs

```diff
@@ -1032,11 +1034,80 @@ impl Binder {
         }
     }
 
+    fn rewrite_format_to_concat_ws(inputs: Vec<ExprImpl>) -> Result<Vec<ExprImpl>> {
+        let Some((format_expr, args)) = inputs.split_first() else {
+            return Err(
+                ErrorCode::BindError("Function `format` takes at least 1 arguments (0 given)".to_string()).into(),
+            );
+        };
+        let ExprImpl::Literal(expr_literal) = format_expr else {
+            return Err(
+                ErrorCode::BindError("Function `format` takes a literal string as the first argument".to_string()).into(),
+            );
+        };
+        let Some(ScalarImpl::Utf8(format_str)) = expr_literal.get_data() else {
+            return Err(
+                ErrorCode::BindError("Function `format` takes a literal string as the first argument".to_string()).into(),
+            );
+        };
+        let formatter = Formatter::parse(format_str)
+            .map_err(|err| -> RwError { ErrorCode::BindError(err.to_string()).into() })?;
+
+        let specifier_count = formatter
+            .nodes()
+            .iter()
+            .filter(|f_node| matches!(f_node, FormatterNode::Specifier(_)))
+            .count();
+        if specifier_count != args.len() {
+            return Err(ErrorCode::BindError(format!(
+                "Function `format` required {} arguments based on the `formatstr`, but {} found.",
+                specifier_count + 1,
+                args.len() + 1
+            ))
+            .into());
+        }
+
+        // iter the args.
+        let mut j = 0;
+        let new_args = [Ok(ExprImpl::literal_varchar("".to_string()))]
+            .into_iter()
+            .chain(formatter.nodes().iter().map(move |f_node| -> Result<_> {
+                let new_arg = match f_node {
+                    FormatterNode::Specifier(sp) => {
+                        // We've checked the count.
+                        let arg = &args[j];
+                        j += 1;
+                        match sp.ty {
+                            SpecifierType::SimpleString => arg.clone(),
+                            SpecifierType::SqlIdentifier => {
+                                FunctionCall::new(ExprType::QuoteIdent, vec![arg.clone()])?.into()
```

Comment:
```suggestion
                                FunctionCall::new(ExprType::QuoteIdent, vec![arg.clone().cast_output()?])?.into()
```

`quote_ident` only handles string input, while `concat_ws` and `format` handles `any`.

> Each *`formatarg`* argument is converted to text according to the usual output rules for its data type, and then formatted and inserted into the result string according to the format specifier(s).

Let's also add some tests for non-string arguments.